### ISSUE TITLE
KNPE in Devirtualization$…$.processVirtualCall

### ISSUE THREAD

**Issue by fluidsonic** [JetBrains/kotlin-native#3087]
Gradle task `linkDebugFrameworkIosArm64` (`iosX64` multiplatform target)

 * Compiler version info: Konan: 1.3 / Kotlin: 1.3.40
 * Output kind: FRAMEWORK

This is in a project which I cannot share and I can't provide a repro project yet because I keep running into [KT-32092](https://youtrack.jetbrains.com/issue/KT-32092) when trying so.

Worked fine with Kotlin 1.3.31.

```
e: kotlin.KotlinNullPointerException
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.processVirtualCall(Devirtualization.kt:276)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.checkSupertypes(Devirtualization.kt:305)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.checkSupertypes(Devirtualization.kt:312)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.addInstantiatingClass(Devirtualization.kt:261)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.dfs(Devirtualization.kt:343)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.processVirtualCall(Devirtualization.kt:280)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.dfs(Devirtualization.kt:396)
        // …
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.dfs(Devirtualization.kt:396)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.search(Devirtualization.kt:251)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis.analyze(Devirtualization.kt:459)
        at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization.run(Devirtualization.kt:1007)
        at org.jetbrains.kotlin.backend.konan.llvm.BitcodePhasesKt$devirtualizationPhase$1.invoke(BitcodePhases.kt:68)
        at org.jetbrains.kotlin.backend.konan.llvm.BitcodePhasesKt$devirtualizationPhase$1.invoke(BitcodePhases.kt)
        at org.jetbrains.kotlin.backend.konan.KonanLoweringPhasesKt$makeKonanModuleOpPhase$1.invoke(KonanLoweringPhases.kt:54)
        at org.jetbrains.kotlin.backend.konan.KonanLoweringPhasesKt$makeKonanModuleOpPhase$1.invoke(KonanLoweringPhases.kt:52)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
        at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
        at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
        at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
        at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
        at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
        at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
        at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
        at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
        at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:41)
        at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:27)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:79)
        at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:34)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:84)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:42)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:104)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:82)
        at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:50)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:215)
        at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:207)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:215)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:212)
        at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:40)
        at org.jetbrains.kotlin.konan.util.UtilKt.profile(Util.kt:34)
        at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:214)
        at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:296)
        at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:16)
```

**Comment by homuroll** [comment]
Seems like one of the dependencies were not rebuilt with the new compiler. Rebuilding them should help.

**Comment by fluidsonic** [comment]
I had to rebuild all dependencies anyway because otherwise I'd get a ABI version mismatch.

I've made a [repro project](https://github.com/fluidsonic/KN-3087) now but the exception has changed there from KNPE to IOOBE.
There are no dependencies except kotlinx-serialization 0.11.1.

```shell
./gradlew clean linkDebugFrameworkIosX64
```
```
 * Source files: Dummy.kt
 * Compiler version info: Konan: 1.3 / Kotlin: 1.3.40
 * Output kind: FRAMEWORK

e: java.lang.IndexOutOfBoundsException: Index: 2, Size: 0
  at java.util.ArrayList.rangeCheck(ArrayList.java:657)
  at java.util.ArrayList.get(ArrayList.java:433)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.processVirtualCall(Devirtualization.kt:273)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.checkSupertypes(Devirtualization.kt:305)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.checkSupertypes(Devirtualization.kt:312)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.addInstantiatingClass(Devirtualization.kt:261)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.dfs(Devirtualization.kt:343)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.processVirtualCall(Devirtualization.kt:280)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.dfs(Devirtualization.kt:396)
  // …
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.dfs(Devirtualization.kt:396)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis$InstantiationsSearcher.search(Devirtualization.kt:251)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization$DevirtualizationAnalysis.analyze(Devirtualization.kt:459)
  at org.jetbrains.kotlin.backend.konan.optimizations.Devirtualization.run(Devirtualization.kt:1007)
  at org.jetbrains.kotlin.backend.konan.llvm.BitcodePhasesKt$devirtualizationPhase$1.invoke(BitcodePhases.kt:68)
  at org.jetbrains.kotlin.backend.konan.llvm.BitcodePhasesKt$devirtualizationPhase$1.invoke(BitcodePhases.kt)
  at org.jetbrains.kotlin.backend.konan.KonanLoweringPhasesKt$makeKonanModuleOpPhase$1.invoke(KonanLoweringPhases.kt:54)
  at org.jetbrains.kotlin.backend.konan.KonanLoweringPhasesKt$makeKonanModuleOpPhase$1.invoke(KonanLoweringPhases.kt:52)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
  at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
  at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
  at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
  at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
  at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
  at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:113)
  at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:23)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:112)
  at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:92)
  at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:41)
  at org.jetbrains.kotlin.backend.konan.KonanDriverKt.runTopLevelPhases(KonanDriver.kt:27)
  at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:79)
  at org.jetbrains.kotlin.cli.bc.K2Native.doExecute(K2Native.kt:34)
  at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:84)
  at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:42)
  at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:104)
  at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:82)
  at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:50)
  at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:215)
  at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:207)
  at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:215)
  at org.jetbrains.kotlin.cli.bc.K2Native$Companion$main$1.invoke(K2Native.kt:212)
  at org.jetbrains.kotlin.konan.util.UtilKt.profileIf(Util.kt:40)
  at org.jetbrains.kotlin.konan.util.UtilKt.profile(Util.kt:34)
  at org.jetbrains.kotlin.cli.bc.K2Native$Companion.main(K2Native.kt:214)
  at org.jetbrains.kotlin.cli.bc.K2NativeKt.main(K2Native.kt:296)
  at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:16)

> Task :framework:linkDebugFrameworkIosX64 FAILED
```

**Comment by CherryPerry** [comment]
Got same issue with our project. We have no external dependencies except kotlin stdlib. Reproducible on my fork/branch https://github.com/CherryPerry/Reaktive/tree/kotlin-1.3.40

Trying `clean cleanBuildCache iosTest --refresh-dependencies --rerun-tasks` from IDEA Gradle Plugin.

**Comment by CherryPerry** [comment]
`-Xdisable-phases=Devirtualization` fixes issue.

**Comment by fluidsonic** [comment]
Repro:
```kotlin
import kotlinx.serialization.*

@Serializable
sealed class Parent {

   @Serializable
   class Child(val foo: String) : Parent()

   companion object {
      val fail = Child.serializer()
   }
}
```

The problem is that `Child.serializer()` is accessed before its companion has been initialized.
Causes an NPE at runtime in the JVM and KNPE or IOOBE in Kotlin Native at compile-time.

**Comment by SvyatoslavScherbina** [comment]
Thanks, confirmed.

**Comment by fluidsonic** [comment]
`-Xdisable-phases=Devirtualization` doesn't seem to help for release builds. Is it possible that devirtualization is run for release builds independent of the presence of the flag?

**Comment by SvyatoslavScherbina** [comment]
> Is it possible that devirtualization is run for release builds independent of the presence of the flag?

Thanks, this appears to be true in 1.3.41. Will be fixed in 1.3.50 with #3223.

In 1.3.50 disabling virtualization requires disabling `DCEPhase` too, i.e. `-Xdisable-phases=Devirtualization,DCEPhase` flag is required.

**Comment by homuroll** [comment]
Another workaround is to forbid serialization of the `Parent` class - it does not make any sense to deserialize sealed (and therefore abstract) class.

**Comment by fluidsonic** [comment]
It does makes sense to deserialize an abstract class.
The serializer can then decide what subclass to instantiate based on the serialized data.
It just cannot be auto-generated.

**Comment by homuroll** [comment]
A, ok, I did mean it's invalid to instantiate sealed class. Anyhow, there will be a fix for sealed classes in the `kotlinx.serialization` library in the upcoming 1.3.50 release.

**Comment by SvyatoslavScherbina** [comment]
Fixed in 1.3.50.
